### PR TITLE
Integration Test case for adaptive sampling

### PR DESCRIPTION
**Pull request by AnmolxSingh**
## Which problem is this PR solving?
Part of  Create e2e integration tests for Adaptive Sampling  #5717 

## Description of the changes
Added test case for adaptive sampling in `adaptive-sampling-integration-test.sh`

## How was this change tested?
currently does not run

### CODE REVIEW

**Comment by yurishkuro** [comment]
please describe the approach first, how are you planning to test changes in the adaptive sampling-calculated probabilities?

**Comment by AnmolxSingh** [comment]
@yurishkuro my approach to testing changes in adaptive sampling-calculated probabilities focuses on validating how Jaeger adjusts its sampling rates based on dynamically calculated probabilities in various operational scenarios. I've set up automated tests using Bash scripts.

Starting Jaeger:

Running Trace Generator: 

Checking Sampling Strategy:

**Comment by yurishkuro** [comment]
please see updated ticket description.

**Comment by AnmolxSingh** [comment]
@yurishkuro 
My approach for adaptive sampling is :

Set default sampling probability to 100%.
Target adaptive sampling at 1 trace per second.

Load Simulation:
Generate 10 traces per second using a trace generator.

Verification:
Record the initial sampling probability (expected to be 100%).
Apply load and wait for adaptive adjustments.
Record the updated sampling probability (expected to decrease, e.g., to around 10% or 50%). 

As per updated ticket description

File: scripts/adaptive-sampling-integration-test.sh [hunk 1]
```diff
@@ -2,16 +2,53 @@
 
 set -euf -o pipefail
 
-# This script is currently a placeholder.
+# Function to start Jaeger with adaptive sampling
+start_jaeger() {
+  echo "Starting Jaeger with adaptive sampling..."
+  SAMPLING_STORAGE_TYPE=memory SAMPLING_CONFIG_TYPE=adaptive go run -tags=ui ../cmd/all-in-one --log-level=debug
+  # Wait for Jaeger to start (adjust sleep duration as needed)
+  sleep 10 &
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:10]: we do not use unconditional sleep in the tests, please see how other scripts are handling readiness or condition checking

File: scripts/adaptive-sampling-integration-test.sh [hunk 2]
```diff
@@ -2,16 +2,53 @@
...
-# - SDK does not report sampling probability in the tags the way Jaeger SDKs did
-# - Server probably does not recognize spans as having adaptive sampling without sampler info
-# - There is no way to modify target traces-per-second dynamically, must restart collector.
+  # Wait for some time to allow for adaptive sampling adjustments (adjust sleep duration as needed)
+  sleep 300
+
+  echo "Updated adaptive sampling strategy:"
+  curl -s 'http://localhost:14268/api/sampling?service=tracegen' | jq .
+
+  initial_prob=$(echo "$initial_strategy" | jq -r '.probabilities[0].probability')
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:32]: `initial_strategy` is never initialized

**Comment by AnmolxSingh** [comment]
@yurishkuro I believe some of recommendations have been pushed. Please review

File: scripts/adaptive-sampling-integration-test.sh [hunk 3]
```diff
@@ -2,16 +2,69 @@
 
 set -euf -o pipefail
 
-# This script is currently a placeholder.
+# Function to start Jaeger with adaptive sampling
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:5]: comment is redundant

File: scripts/adaptive-sampling-integration-test.sh [hunk 4]
```diff
@@ -2,16 +2,69 @@
...
+start_jaeger() {
+  echo "Starting Jaeger with adaptive sampling..."
+  SAMPLING_STORAGE_TYPE=memory SAMPLING_CONFIG_TYPE=adaptive go run -tags=ui ../cmd/all-in-one --log-level=debug
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:8]: you are not passing the desired initial sampling rate, as you mentioned in the plan

File: scripts/adaptive-sampling-integration-test.sh [hunk 5]
```diff
@@ -2,16 +2,69 @@
...
 
-# Commands to run integration test:
-#   SAMPLING_STORAGE_TYPE=memory SAMPLING_CONFIG_TYPE=adaptive go run -tags=ui ./cmd/all-in-one --log-level=debug
-#   go run ./cmd/tracegen -adaptive-sampling=http://localhost:14268/api/sampling -pause=10ms -duration=60m
+  # Wait for Jaeger to start (adjust sleep duration as needed)
+  until curl -s 'http://localhost:16686' > /dev/null; do
+    echo "waiting for Jaeger to start"
+    sleep 2
+  done
+  echo "Jeager started"
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:15]:
```suggestion
  echo "Jaeger started"
```

File: scripts/adaptive-sampling-integration-test.sh [hunk 6]
```diff
@@ -2,16 +2,69 @@
...
+# Function to run trace generator with adaptive sampling
+run_trace_generator() {
+  echo "Running trace generator with adaptive sampling..."
+  go run ./cmd/tracegen -adaptive-sampling=http://localhost:14268/api/sampling -pause=10ms -duration=60m
 
-# Issues
-# - SDK does not report sampling probability in the tags the way Jaeger SDKs did
-# - Server probably does not recognize spans as having adaptive sampling without sampler info
-# - There is no way to modify target traces-per-second dynamically, must restart collector.
+  # Wait for trace generation to complete (adjust duration as needed)
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:23]: why do you need to wait? It's set to run for 60min. The point was to create a sustained level of load and check how it affect computed probability. Then kill the load generator (not wait till it's done).

File: scripts/adaptive-sampling-integration-test.sh [hunk 7]
```diff
@@ -2,16 +2,69 @@
...
+check_sampling_strategy() {
+  echo "Initial adaptive sampling strategy:"
+  initial_strategy=$(curl -s 'http://localhost:14268/api/sampling?service=tracegen')
+  echo "Initial adaptive sampling stratergy:"
+  echo "$initial_strategy" | jq.
+  
+  # Wait for some time to allow for adaptive sampling adjustments (adjust sleep duration as needed)
+  adjustment_duration=30
+  echo "Waiting for $adjustment_duration seconds to allow for adaptive sampling adjustments..."
+  sleep $adjustment_duration
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:41]: Once again, we don't just sleep for random durations. CI runs on shared VMs, it's unpredictable how much CPU quota a given runner might get (or get starved of). Sleeps are only used in small increments when checking for a specific condition in a loop.

**Review by yurishkuro — COMMENTED** [review]
I am quite certain the script does not work at all, as you're missing some important orchestration mechanisms. But I don't want to point them out, because I expect that if you submit some PR you actually have tested it locally and verified that it works (I think this one would just hang forever). Sometimes it's ok to submit a PR if you're asking for an opinion on the approach / direction, but you didn't seem to ask that, you are presenting this as a solution.

**Comment by AnmolxSingh** [comment]
@yurishkuro 
Thank you for your feedback. I apologize if my message came across as presenting the script as a finalized solution. As I am an absolute beginner to this project, I am actually seeking your opinion on the approach and direction of my work so far. I am still learning and trying to understand the orchestration mechanisms needed for this task.

I appreciate your guidance and any advice you can offer to help me improve and correctly implement the necessary changes.

File: scripts/adaptive-sampling-integration-test.sh [hunk 8]
```diff
@@ -2,16 +2,92 @@
 
 set -euf -o pipefail
 
-# This script is currently a placeholder.
+start_jaeger() {
+  echo "Starting Jaeger with adaptive sampling..."
+  SAMPLING_STORAGE_TYPE=memory SAMPLING_CONFIG_TYPE=adaptive go run -tags=ui ../cmd/all-in-one --log-level=debug &
```
Comment (yurishkuro) [inline scripts/adaptive-sampling-integration-test.sh:7]: you need to clean up all background processes once the script exists (successfully or with error). Usually we use `trap` for that. I would recommend capturing PIDs into global variables and calling a function via `trap` that will will `kill -9` on all those PIDs (those that are not empty).

**Review by yurishkuro — COMMENTED** [review]
> I am actually seeking your opinion on the approach and direction of my work so far. 

Your direction is ok at this point. If you have specific questions on direction, you should ask questions in the comments. But there is no point submitting code that just doesn't work, test it locally instead. If you run into an issue where something doesn't work, then ask questions.

**Pull request closed without merging**